### ISSUE TITLE
`reportUninitializedInstanceVariable` false positive inside a dataclass' `__post_init__`

### ISSUE THREAD

**Issue by tpvasconcelos** [microsoft/pyright#9623]
**`reportUninitializedInstanceVariable` false positive when declaring new instance variables inside a dataclass' `__post_init__` method.**

Defining derived instance variables inside a `__post_init__` method is quite a common usage pattern when defining new dataclasses. Take the following example:

```python
from dataclasses import dataclass

@dataclass
class Person:
    first_name: str
    last_name: str
    
    def __post_init__(self) -> None:
        self.full_name = f"{self.first_name} {self.last_name}"  # pyright: ignore[reportUninitializedInstanceVariable]
```

There are some workarounds to make pyright happy (such as turning this into a `@property`), but I don't think there's a valid reason for pyright not accepting the code above since `full_name` will be a valid attribute after instantiation.

```python
jon = Person("Jon", "Doe")
jon.full_name  # 'Jon Doe'
```

**VS Code extension or command-line**

Running both the pyright CLI and language server in PyCharm (version `1.1.391`)

**Comment by erictraut** [comment]
The `reportUninitializedInstanceVariable` check is a highly-opinionated check that is not enabled by default even in `strict` mode. It requires you to follow very specific rules — namely, that the instance variable needs to be initialized in either the class body or within an `__init__` method. It doesn't currently offer special-casing for other initialization methods like `__post_init__`. If you want to use this check, you'll need to either conform to the limitations of the check or work around cases where these limitations are too constraining for your coding style.

**Comment by tpvasconcelos** [comment]
@erictraut Thanks for the super quick reply! (also thanks for maintaining this great tool!)

Are there any plans to every support this? I understand that supporting multiple special cases can be unfeasible but since dataclasses are part of Python's standard-lib, I wonder if an exception can be made here....

**Comment by erictraut** [comment]
There are currently no plans to add such special-casing. I'm willing to consider it if there's sufficient signal from other pyright users (e.g. in the form of upvotes to this issue).
